### PR TITLE
feat(cli): EAS build hooks & asyncUpload

### PR DESCRIPTION
**Pull request by michalziolkowski**
This PR provides support for asyncUpload and EAS builds.

### 1. To initialize build without passing android/ios files, add `asyncUpload` option to:

**CLI**
```
npx @sherlo/cli --asyncUpload
```

**Github Action**
```
- name: Run Sherlo Action with Async Upload
  id: sherlo_action # This ID is used to reference the output
  uses: sherlo-io/sherlo@latest
  with:
    asyncUpload: true
  env:
    SHERLO_TOKEN: ${{ secrets.SHERLO_TOKEN }}
 ```
 
### 2. You will get buildIndex in output

**CLI**
`...Build index is ${output.buildIndex}.`

**Github Action**
`steps.sherlo_action.outputs.buildIndex` variable will contain the buildIndex of created build

### 3. Once you have a native build you can send it using `asyncUploadBuildIndex`

**CLI**
```
npx @sherlo/cli --asyncUploadBuildIndex=1 --android="build/android.apk"
```

**Github Action**
```
- name: Upload Android to Sherlo Asynchronously
  uses: sherlo-io/sherlo@latest
  with:
    asyncUploadBuildIndex: ${{ steps.sherlo_action.outputs.buildIndex }}
    android:"build/android.apk"
  env:
    SHERLO_TOKEN: ${{ secrets.SHERLO_TOKEN }}
 ```
 
## EAS hooks

Support for EAS builds is based on asyncUpload. User follows the first step in the same manner

### 1. Add "eas-build-on-success" script to your `package.json`

`"eas-build-on-success": "sherlo-eas-build-on-success <PROFILE_OF_SIM_EAS_BUILD>",`

and make sure you have `@sherlo/cli` added to `devDependencies`

### 2. Start Sherlo with `asyncUpload`

You need to make sure to run in on the same copy of the repo you will use for EAS build as Sherlo will create `.expo/sherlo.json` file to store `buildIndex` value that will be used later

GH Action example, but you can use CLI locally too:
```
# It's important to run Sherlo in the same job as the EAS build,
      # because we're pushing temporary files to .expo directory used during the build
      - name: Run Sherlo with Async Upload
        uses: ./.
        with:
          projectRoot: examples/expo-example
          asyncUpload: true
        env:
          SHERLO_TOKEN: ${{ secrets.SHERLO_TOKEN_EAS }}

      - name: Trigger EAS builds for both platforms
        working-directory: examples/expo-example
        run: eas build --profile=<PROFILE_OF_SIM_EAS_BUILD> --platform=all --non-interactive --no-wait
 ```
 
 ### 3. Once the build is finished it will trigger the `eas-build-on-success` hook on expo servers and will upload the build to sherlo. Once all builds are uploaded Sherlo starts tests automatically
 
## Testing workflows

This PR also introduces automated workflows to test various sherlo use cases for upload:
- `Run Sherlo with Sync Upload` - this can be triggered on PR with `sherlo:sync` and `sherlo:sync:skip_build` labels
- `Run Sherlo with Async Upload` - this can be triggered on PR with `sherlo:async` and `sherlo:async:skip_build` labels
- `Run Sherlo with EAS hooks` - this can be triggered on PR with `sherlo:eas` label
 
 They also can be run as a part of manually triggered `Sherlo Tests` workflow where you can select branch and set of tests you want to run - this will be useful for testing before releases
 
 I was able to test these [workflows](https://github.com/sherlo-io/sherlo/actions/runs/8443964142) by making some code changes but these will only be runnable once we merge this pr
 
image

### CODE REVIEW

File: packages/cli/eas-hooks/eas-build-on-success.sh [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+if [ -z "$SHERLO_TOKEN" ]; then
+    echo "Error: You must provide valid sherlo project token under SHERLO_TOKEN secret variable in Expo."
+    exit 1
+fi
+
+SHERLO_BUILD_PROFILE=$1
+
+if [ -z "$SHERLO_BUILD_PROFILE" ]; then
+    echo "Error: You must provide the EAS profile as an argument to this script."
```
Comment (michalziolkowski) [inline packages/cli/eas-hooks/eas-build-on-success.sh]: dodac info ze to musi byc nazwa profilu ktory chcesz wysylac do sherlo na test

File: packages/cli/src/commands/main/_getArguments.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  if (args.config) args.config = path.join(args.projectRoot, args.config);
+
+  // parse config
+  const config = parse(args.config);
+
+  // adjust config paths for projectRoot path
+  if (config.android?.path) config.android.path = path.join(args.projectRoot, config.android?.path);
+  if (config.ios?.path) config.ios.path = path.join(args.projectRoot, config.ios?.path);
+
+  // override config paths paths with CLI / GH action args
```
Comment (michalziolkowski) [inline packages/cli/src/commands/main/_getArguments.ts]: typo

File: packages/cli/src/commands/main/_getArguments.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+  if (args.android && config.android) config.android.path = args.android;
+  if (args.ios && config.ios) config.ios.path = args.ios;
+
+  const token = args.token || config.token;
```
Comment (michalziolkowski) [inline packages/cli/src/commands/main/_getArguments.ts]: ??

File: packages/cli/src/commands/main/main.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        teamId,
+        projectIndex,
+        buildRunConfig: getBuildRunConfig({ config: args.config }),
+        asyncUpload: true,
+        gitInfo: args.gitInfo,
+      });
+
+      const output = createOutput({ buildIndex: build.index, projectIndex, teamId });
+
+      const expoDir = path.join(args.projectRoot, '.expo');
```
Comment (michalziolkowski) [inline packages/cli/src/commands/main/main.ts]: sprawdzic czy w `package.json` jest skrypt "eas-build-on-success"

**Review by dawidk92 — APPROVED** [review]

**Pull request merged**